ompi/mca/coll/han/coll_han_alltoallv.c

```diff
@@ -0,0 +1,878 @@
+        avg_send_bytes += packed_size_bytes/count_for_convertor * ompi_count_array_get(scounts,jrank);
+    }
+    need_bufs |= opal_convertor_need_buffers(&convertor);
+
+    rc |= opal_convertor_cleanup(&convertor);
+    OBJ_DESTRUCT( &convertor );
+
+    if (rc != OMPI_SUCCESS) { return rc;}
+    avg_send_bytes = avg_send_bytes / comm_size;
+    reduce_buf_input[0] = !!(bufs_on_device);
+    reduce_buf_input[1] = avg_send_bytes;
+    reduce_buf_input[2] = !!(need_bufs);
+    rc =comm->c_coll->coll_allreduce(
+        &reduce_buf_input, &reduce_buf_output, 3, MPI_LONG_LONG, &ompi_mpi_op_sum.op,
+        comm, comm->c_coll->coll_allreduce_module );
+    if (rc != OMPI_SUCCESS) {return rc;}
+
+    if (reduce_buf_output[0] > 0) {
+        *use_smsc = 0;
+        /* can't proceed: at least one rank using GPU buffers. */
+    } else if (reduce_buf_output[2] >= comm_size * mca_coll_han_component.han_alltoallv_smsc_noncontig_activation_limit) {
+        /* always proceed: enough ranks have non-contiguous data that the pack-and-send method is fastest. */
+        *use_smsc = 1;
+    } else if (reduce_buf_output[1] >= comm_size * mca_coll_han_component.han_alltoallv_smsc_avg_send_limit) {
+        *use_smsc = 0;
+        /* don't proceed: messages are large and contiguous.  It is faster to fall back to basic alg. */
+    } else {
+        *use_smsc = 1;
+    }
+
+    if (comm_rank == 0) {
+        opal_output_verbose(10, mca_coll_han_component.han_output, "alltoallv: decide_to_use_smsc_alg: "
+            "Ranks with GPU buffers: %lld (limit is 0).  "
+            "Average send_size: %.1f bytes (limit is %ld).  "
+            "Fraction with non-contiguous buffers: %.3f (activation limit: %.3f).  "
+            "Continue with SMSC? ==>%s.\n",
+            reduce_buf_output[0],
+            ((double)reduce_buf_output[1])/comm_size, mca_coll_han_component.han_alltoallv_smsc_avg_send_limit,
+            ((double)reduce_buf_output[2])/comm_size, mca_coll_han_component.han_alltoallv_smsc_noncontig_activation_limit,
+            (*use_smsc) ? "Yes" : "No");
+    }
+    return rc;
+}
+
+int mca_coll_han_alltoallv_using_smsc(
+        const void *sbuf,
+        ompi_count_array_t scounts,
+        ompi_disp_array_t sdispls,
+        struct ompi_datatype_t *sdtype,
+        void* rbuf,
+        ompi_count_array_t rcounts,
+        ompi_disp_array_t rdispls,
+        struct ompi_datatype_t *rdtype,
+        struct ompi_communicator_t *comm,
+        mca_coll_base_module_t *module)
+{
+
+    OPAL_OUTPUT_VERBOSE((90, mca_coll_han_component.han_output,
+                            "Entering mca_coll_han_alltoall_using_smsc\n"));
+    int rc;
+
+    {
+        static int attach_done=0;
+        if (attach_done) goto resume;
+        char *dbg_rank = getenv("DEBUG_RANK");
+        if (!dbg_rank) goto resume;
+        if (atoi(dbg_rank) != ompi_comm_rank(MPI_COMM_WORLD)) goto resume;
+        volatile int i = 0;
+        char hostname[256];
+        sleep(1);
```

Comment:
Sleep(1) ? gethostname ? What's the scope of all these ? Why is this necessary on each call ?

Reply:
Oops.  That's debug code that slipped past me.  I've removed it.